DRIVERS-2630: add e2e testing for search index commands and clarifications to search index spec (#1442)
mongodb/specifications: source/index-management/tests/README.rst

collection name.  This limitation does not take into account collection uuid.  Because these commands are asynchronous, any cleanup
code that may run after a test (cleaning a database or dropping search indexes) may not have completed by the next iteration of the 
test (or the next test run, if running locally).  To address this issue, each test uses a randomly generated collection name.  Drivers
may generate this collection name however they like, but a suggested implementation is a hex representation of an
ObjectId (``new ObjectId().toHexString()`` in Node).

Setup
~~~~~

These tests must run against an Atlas cluster with a 7.0+ server.  Tools are available drivers-evergreen-tools which can setup and teardown
Atlas clusters.  To ensure that the Atlas cluster is cleaned up after each CI run, drivers should configure evergreen to run these tests 
as a part of a task group.  Be sure that the cluster gets torn down! 

When working locally on these tests, the same Atlas setup and teardown scripts can be used locally to provision a cluster for development.

Case 1: Driver can successfully create and list search indexes
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> #. Create a collection with a randomly generated name (referred to as ``coll0``).

Reviewer: I confirm the collection must be created first.

#. Create a new search index on ``coll0`` with the ``createSearchIndex`` helper.  Use the following definition:

  .. code:: typescript

    {
      name: 'test-search-index',
      definition: {
        mappings: { dynamic: false }
      }
    }

#. Assert that the command returns the name of the index: ``"test-search-index"``.
#. Run ``coll0.listSearchIndexes()`` repeatedly every 5 seconds until the following condition is satisfied:
   1. An index with the ``name`` of ``test-search-index`` is present and the index has a field ``queryable`` with a value of ``true``.

> #. Assert that ``index`` has a property ``mappings`` whose value is ``{ dynamic: false }``

Reviewer 2: Should we specify on line 95 that the result of `coll0.listSearchIndexes()` is stored in a variable called `index`? `index` is referenced on line 98 without any prior mention of initialization. Something similar is done on line 124 where it's mentioned to store the result of a list command into `indexes` (which is later referenced).

Case 2: Driver can successfully create multiple indexes in batch
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

#. Create a collection with a randomly generated name (referred to as ``coll0``).
#. Create two new search indexes on ``coll0`` with the ``createSearchIndexes`` helper.  Use the following
   definitions when creating the indexes.  These definitions are referred to as ``indexDefinitions``.

  .. code:: typescript

    {
      name: 'test-search-index-1',
      definition: {
        mappings: { dynamic: false }
      }
    }
    {
      name: 'test-search-index-2',
      definition: {
        mappings: { dynamic: false }
      }
    }

#. Assert that the command returns an array containing the new indexes' names: ``["test-search-index-1", "test-search-index-2"]``.

> #. Run ``coll0.listSearchIndexes()`` repeatedly every 5 seconds until the following condition is satisfied.  Store
>    the result in ``indexes``.
>    1. An index with the ``name`` of ``test-search-index-1`` is present and index has a field ``queryable`` with the value of ``true``.
>    2. An index with the ``name`` of ``test-search-index-2`` is present and index has a field ``queryable`` with the value of ``true``.
> #. For each ``index`` in ``indexDefinitions``
>    1. Find the matching index definition in ``indexes`` by matching on ``index.name``.  If no index exists, raise an error.
>    2. Assert that the matching index ``mappings``, whose value is ``{ dynamic: false }``

Reviewer 3: ```suggestion
#. Run ``coll0.listSearchIndexes()`` repeatedly every 5 seconds until the following condition is satisfied.
   1. An index with the ``name`` of ``test-search-index-1`` is present and index has a field ``queryable`` with the value of ``true``. Store result in ``index1``.
   2. An index with the ``name`` of ``test-search-index-2`` is present and index has a field ``queryable`` with the value of ``true``. Store result in ``index2``.
#. Assert that ``index1`` and ``index2`` have the property ``latestDefinition`` whose value is ``{ "mappings" : { "dynamic" : false } }``
```

Suggest using separate `index1` and `index2` variables to simplify test.

Suggest updating assert to use `latestDefinition` to match returned document. Here is an example of the documents returned in the C driver testing:

```
  { "id" : "64c15fc9a199d3199e27ad70", "name" : "test-search-index-1", "status" : "READY", "queryable" : true, "latestDefinition" : { "mappings" : { "dynamic" : false } } }
  { "id" : "64c15fc9a199d3199e27ad71", "name" : "test-search-index-2", "status" : "READY", "queryable" : true, "latestDefinition" : { "mappings" : { "dynamic" : false } } }
```

Case 3: Driver can successfully drop search indexes
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

#. Create a collection with a randomly generated name (referred to as ``coll0``).
#. Create a new search index on ``coll0`` with the following definition:

  .. code:: typescript

    {
      name: 'test-search-index',
      definition: {
        mappings: { dynamic: false }
      }
    }

#. Assert that the command returns the name of the index: ``"test-search-index"``.
#. Run ``coll0.listSearchIndexes()`` repeatedly every 5 seconds until the following condition is satisfied:
   1. An index with the ``name`` of ``test-search-index`` is present and index has a field ``queryable`` with the value of ``true``.

> #. Run a ``dropSearchIndexes`` on ``coll0``, using ``test-search-index`` for the name.

Reviewer 3: ```suggestion
#. Run a ``dropSearchIndex`` on ``coll0``, using ``test-search-index`` for the name.
```

  .. code:: typescript

    {
      name: 'test-search-index',
      definition: {
        mappings: { dynamic: false }
      }
    }

#. Assert that the command returns the name of the index: ``"test-search-index"``.
#. Run ``coll0.listSearchIndexes()`` repeatedly every 5 seconds until the following condition is satisfied:
   1. An index with the ``name`` of ``test-search-index`` is present and index has a field ``queryable`` with the value of ``true``.

#. Run a ``updateSearchIndex`` on ``coll0``, using the following definition.
  
  .. code:: typescript

    {
      name: 'test-search-index',
      definition: {
        mappings: { dynamic: true }
      }
    }

#. Assert that the command does not error and the server responds with a success.
#. Run ``coll0.listSearchIndexes()`` repeatedly every 5 seconds until the following condition is satisfied:
   1. An index with the ``name`` of ``test-search-index`` is present.  This index is referred to as ``index``.
   2. The index has a field ``queryable`` with a value of ``true`` and has a field ``status`` with the value of ``READY``.
  
#. Assert that an index is present with the name ``test-search-index`` and the definition has a

>   property ``mappings``, whose value is ``{ dynamic: true }``

Reviewer 3: ```suggestion
  property ``latestDefinition``, whose value is ``{ 'mappings' : { 'dynamic' : true } }``
```

Suggest updating assert to use `latestDefinition` to match returned document. Here is an example of the documents returned in the C driver testing:

```
{ "id" : "64c1651aa199d3199e27ae1a", "name" : "test-search-index", "status" : "PENDING", "queryable" : false, "latestDefinition" : { "mappings" : { "dynamic" : false } } }
```

test (or the next test run, if running locally).  To address this issue, each test uses a randomly generated collection name.  Drivers
may generate this collection name however they like, but a suggested implementation is a hex representation of an
ObjectId (``new ObjectId().toHexString()`` in Node).

Setup
~~~~~

These tests must run against an Atlas cluster with a 7.0+ server.  Tools are available drivers-evergreen-tools which can setup and teardown
Atlas clusters.  To ensure that the Atlas cluster is cleaned up after each CI run, drivers should configure evergreen to run these tests 
as a part of a task group.  Be sure that the cluster gets torn down! 

When working locally on these tests, the same Atlas setup and teardown scripts can be used locally to provision a cluster for development.

Case 1: Driver can successfully create and list search indexes
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

#. Create a collection with a randomly generated name (referred to as ``coll0``).
#. Create a new search index on ``coll0`` with the following definition:

  .. code:: typescript

    {

>       name: 'test-search-index',

Reviewer: Correct, the server may decide to change the default name in a future version. We don't have to test it. And the UST test the command sent when no name is provided.

      }
    }

#. Assert that the command does not error and the server responds with a success.
#. Run ``coll0.listSearchIndexes()`` repeatedly every 5 seconds until the following condition is satisfied:
   1. An index with the ``name`` of ``test-search-index`` is present.  This index is referred to as ``index``.
   2. The ``latestDefinition`` property of the search index has the field ``queryable`` with a value of ``true``.
#. Assert that ``index`` has a property ``mappings`` whose value is ``{ dynamic: false }``

Case 2: Driver can successfully create multiple indexes in batch
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

#. Create a collection with a randomly generated name (referred to as ``coll0``).
#. Create two new search indexes on ``coll0`` with the ``createSearchIndexes`` command.  Use the following
   definitions when creating the indexes.  These definitions are referred to as ``indexDefinitions``.

  .. code:: typescript

    {
      name: 'test-search-index-1',
      definition: {
        mappings: { dynamic: false }
      }
    }
    {
      name: 'test-search-index-2',
      definition: {
        mappings: { dynamic: false }
      }
    }

> #. Assert that the command does not error and the server responds with a success.

Author: Done!

> #. Run ``coll0.listSearchIndexes()`` repeatedly every 5 seconds until the following condition is satisfied.  Store

Reviewer: If the driver sends `{ name: null }`, the servers produces this error: `PlanExecutor error during aggregation :: caused by :: org.bson.BsonInvalidOperationException: Value expected to be of type STRING is of unexpected type NULL`

But I see this is already covered by the UST.

   1. An index with the ``name`` of ``test-search-index`` is present.  This index is referred to as ``index``.
   2. The ``latestDefinition`` property of the search index has the field ``queryable`` with a value of ``true``.

#. Run a ``updateSearchIndex`` on ``coll0``, using the following definition.
  
  .. code:: typescript

    {
      name: 'test-search-index',
      definition: {
        mappings: { dynamic: true }
      }
    }

#. Assert that the command does not error and the server responds with a success.
#. Run ``coll0.listSearchIndexes()`` repeatedly every 5 seconds until the following condition is satisfied:
   1. An index with the ``name`` of ``test-search-index`` is present.  This index is referred to as ``index``.
   2. The ``latestDefinition`` property of the search index has the field ``queryable`` with a value of ``true``.
   3. The ``latestDefinition`` property of the search index has the field ``status`` with the value of ``READY``.
  
#. Assert that an index is present with the name ``test-search-index`` and the definition has a
  property ``mappings``, whose value is ``{ dynamic: true }``

> Case 5: ``dropSearchIndex`` suppresses namespace not found errors

Reviewer: I didn't see this in the spec. I'm ignoring [error with code `26`](https://github.com/mongodb/mongo/blob/master/src/mongo/base/error_codes.yml)

Reviewer: Nothing to change. It's just me who didn't read the spec as carefully as I should have.

It's [in the spec](https://github.com/mongodb/specifications/blob/78554de6a78fc996fa8d1ef9733b123a29ad8309/source/index-management/index-management.rst#namespacenotfound-errors).

> Drivers MUST suppress NamespaceNotFound errors for the dropSearchIndex helper.
